### PR TITLE
draft schema design

### PR DESCRIPTION
**Pull request by Vidushi-GitHub**
[#921](https://github.com/nasa-gcn/gcn.nasa.gov/issues/921) in progress

### CODE REVIEW

**Comment by dakota002** [comment]
A useful link to have somewhere in this documentation: https://www.learnjsonschema.com/2020-12/

**Comment by Vidushi-GitHub** [comment]
Add schema related step at New Notice Producers.

File: app/routes/docs/producers/schema-documentation.tsx [hunk 1]
```diff
@@ -0,0 +1,158 @@
+// ---
```
Comment (lpsinger) [inline app/routes/docs/producers/schema-documentation.tsx:1]: Please convert this file to .mdx.

**Review by lpsinger — CHANGES_REQUESTED** [review]

**Comment by Vidushi-GitHub** [comment]
Attached transition from .tsx to .mdx is still in process
sample_code_mdx
. Suggestions are welcomed, I couldn't locate eg in other .mdx files here.

**Comment by Vidushi-GitHub** [comment]
For Schema Practices page the draft is attached
Screenshot 2023-05-26 at 4 16 33 PM
, we can keep adding questions.

**Comment by Vidushi-GitHub** [comment]
A step has been added at New Notice Producers page. Please the screenshot @jracusin:
Screenshot 2023-05-31 at 3 32 21 PM

**Comment by Vidushi-GitHub** [comment]
The initial schema design, please see the screenshots for the steps:
Screenshot 2023-05-31 at 3 34 41 PM
d-b90b-48a7-a82a-00d58c401e78">

**Comment by Vidushi-GitHub** [comment]
@dakota002 kindly make the requested changes as discussed.
Screenshot 2023-05-31 at 3 50 47 PM

**Comment by Vidushi-GitHub** [comment]
@jracusin and @lpsinger schema design page is done from our side (@dakota002). I am not in favor of adding Alert.Schema example of Guano, looks redundant, as already present at Schema Browser and Git repo. But Dakota finds it helpful. Your suggestions would be helpful.

Screenshot 2023-06-01 at 7 11 03 PM
Screenshot 2023-06-01 at 7 11 17 PM
Screenshot 2023-06-01 at 7 11 31 PM
Screenshot 2023-06-01 at 7 11 53 PM
Screenshot 2023-06-01 at 7 12 00 PM

**Comment by jracusin** [comment]
For submitting schema, we need to add something about making a PR with their branch, that's what initiates our review.  Also do we really need the details of exactly how to use git?

**Comment by Vidushi-GitHub** [comment]
> For submitting schema, we need to add something about making a PR with their branch, that's what initiates our review. Also do we really need the details of exactly how to use git?

Done!

**Comment by Vidushi-GitHub** [comment]
@lpsinger This pull request is ready for review! I am attaching screenshots for reference. Seperate attachments for HealPix page is removed for now, which will require parsing healpix instructions from you and @israelmcmc.
schema_doc1
schema_doc2
tps://github.com/nasa-gcn/gcn.nasa.gov/assets/103223246/0
schema_doc4
827a7c1-240a-4865-aef0-580ee3f4017c">

**Review by lpsinger — CHANGES_REQUESTED** [review]
Please rebase. Sorry, we just merged #840, which touched nearly all of the route modules.

**Comment by Vidushi-GitHub** [comment]
> Please rebase. Sorry, we just merged #840, which touched nearly all of the route modules.

No problem.

File: app/routes/docs.producers.schema-documentation.mdx [hunk 2]
```diff
@@ -0,0 +1,204 @@
+---
```
Comment (lpsinger) [inline app/routes/docs.producers.schema-documentation.mdx:1]: Rename this file to `docs.producers.schema.mdx`. (The `-documentation` bit is redundant, because this is a sub route of docs.)

File: app/routes/docs.faq.md [hunk 3]
```diff
@@ -66,6 +66,30 @@ This warning means that there have not been any recent alerts on that topic.
 
 To get started, [sign in or sign up](https://gcn.nasa.gov/login) and then select 'Email Notifications' from account dropdown menu. See also [GCN Circular 32517](https://gcn.gsfc.nasa.gov/gcn3/32517.gcn3).
 
+## Notice Producers
+
+### Which unit system is recommended to be used?
```
Comment (lpsinger) [inline app/routes/docs.faq.md:71]: Eliminate passive voice.

File: app/routes/docs.faq.md [hunk 4]
```diff
@@ -66,6 +66,30 @@ This warning means that there have not been any recent alerts on that topic.
...
+
+To ensure a consistent database for same property, but different instrument, it is preferable to utilize SI units or maintain the same unit system. Easy conversion can be made using the astropy software.
+
+### Which option, "allOf", "anyOf", or "oneOf" should be used?
+
+We recommend utilizing the "allOf" option, as none of the properties in the core schema are required. However, this flexibility, with no required property, can lead to false positives when "anyOf" is used, as anything will technically validate against it.
+
+### How can I refer the coincident detection with another instrument?
+
+A core schema is created to reference instrument properties for instruments that are conducting follow-up searches coinciding with other missions.
```
Comment (lpsinger) [inline app/routes/docs.faq.md:81]: This doesn't seem to answer the question.

File: app/routes/docs.faq.md [hunk 5]
```diff
@@ -66,6 +66,30 @@ This warning means that there have not been any recent alerts on that topic.
...
+
+### Which unit system is recommended to be used?
+
+To ensure a consistent database for same property, but different instrument, it is preferable to utilize SI units or maintain the same unit system. Easy conversion can be made using the astropy software.
+
+### Which option, "allOf", "anyOf", or "oneOf" should be used?
+
+We recommend utilizing the "allOf" option, as none of the properties in the core schema are required. However, this flexibility, with no required property, can lead to false positives when "anyOf" is used, as anything will technically validate against it.
+
+### How can I refer the coincident detection with another instrument?
```
Comment (lpsinger) [inline app/routes/docs.faq.md:79]: This sentence seems non-grammatical. @jracusin, would you please help re-word it?

File: app/routes/docs.faq.md [hunk 6]
```diff
@@ -66,6 +66,30 @@ This warning means that there have not been any recent alerts on that topic.
...
+
+A core schema is created to reference instrument properties for instruments that are conducting follow-up searches coinciding with other missions.
+
+### How is the schema version tagged?
+
+(ex. Deprecation of fields that are being updated or replaced,)
```
Comment (lpsinger) [inline app/routes/docs.faq.md:85]: This looks like an unfinished section.

File: app/routes/docs.producers._index.mdx [hunk 7]
```diff
@@ -44,6 +44,12 @@ also happy to work with the mission teams to help construct your alerts.
    your list of team members from Step 1 and your chosen Kafka topic prefix from Step 2.
       The GCN Team will reply after they have configured producer permissions for your team.
    </ProcessListItem>
+   <ProcessListItem>
+     <ProcessListHeading type="h3">
+       Draft Your Schema 
+     </ProcessListHeading>
+     As a GCN Notice producer, you have the flexibility to create your own instrument-specific schema. Please contribute your schema to our [Github repository](https://github.com/nasa-gcn/gcn-schema), placing it in a folder under <code>gcn/notices/<i>mission</i></code> and submit a pull request for the GCN Team to review. For details, please refer to the [schema design](/docs/producers/schema-documentation) documentation.
```
Comment (lpsinger) [inline app/routes/docs.producers._index.mdx:51]: GitHub is mis-capitalized:

```suggestion
     As a GCN Notice producer, you have the flexibility to create your own instrument-specific schema. Please contribute your schema to our [GitHub repository](https://github.com/nasa-gcn/gcn-schema), placing it in a folder under <code>gcn/notices/<i>mission</i></code> and submit a pull request for the GCN Team to review. For details, please refer to the [schema design](/docs/producers/schema-documentation) documentation.
```

File: app/routes/docs.producers._index.mdx [hunk 8]
```diff
@@ -60,10 +66,18 @@ also happy to work with the mission teams to help construct your alerts.
       from gcn_kafka import Producer
       # Connect as a producer.
       # Warning: don't share the client secret with others.
-      producer = Producer(client_id='fill me in', client_secret='fill me in')
+      producer = Producer(client_id='fill me in',
+                          client_secret='fill me in')
```
Comment (lpsinger) [inline app/routes/docs.producers._index.mdx:70]: Unnecessary whitespace change.

File: app/routes/docs.producers._index.mdx [hunk 9]
```diff
@@ -60,10 +66,18 @@ also happy to work with the mission teams to help construct your alerts.
...
       # any topic starting with 'mission.'
       topic = 'gcn.notices.mission.example'
-      data = b'...'  # any bytes
+
+      # JSON data converted to byte string format
+      data = str({
+         '$schema':'your/schema',
+         'key':'value'
+         ...
+      }).encode()  # any bytes
```
Comment (lpsinger) [inline app/routes/docs.producers._index.mdx:80]: No, that's not how you serialize a data structure to JSON in Python. Use [`json.dumps`](https://docs.python.org/3/library/json.html#json.dumps).

I don't think that you need to explicitly encode it to bytes. If memory servers, `Producer.produce` happily accepts either strings or bytes. (Please test that.)

**Review by lpsinger — CHANGES_REQUESTED** [review]
The FAQ items are great, but I am not sure that they belong in the FAQ. I think a tutorial or primer on writing JSON Schema would be more useful. Then, your content for the FAQ entry titled "Which option, "allOf", "anyOf", or "oneOf" should be used?" would go into a section of the primer on inheritance.

File: app/routes/docs.producers._index.mdx (hunk 9 above)
Comment (Vidushi-GitHub) [reply app/routes/docs.producers._index.mdx:80]: We will update to use json.dumps

We tested producer.produce with a string and it threw errors when not a byte string

@dakota002 please check this is true

**Comment by Vidushi-GitHub** [comment]
> 

@lpsinger do you suggest that FAQ related to notice producers should be placed under New Notice Producers as Tutorial Section (similar as schema document)? @jracusin @dakota002

File: app/routes/docs.producers._index.mdx [hunk 10]
```diff
@@ -61,9 +67,16 @@ also happy to work with the mission teams to help construct your alerts.
       # Connect as a producer.
       # Warning: don't share the client secret with others.
       producer = Producer(client_id='fill me in', client_secret='fill me in')
+
       # any topic starting with 'mission.'
       topic = 'gcn.notices.mission.example'
-      data = b'...'  # any bytes
+
+      # JSON data converted to byte string format
+      data = str({
```
Comment (Vidushi-GitHub) [inline app/routes/docs.producers._index.mdx:75]:
```suggestion
      data = json.dumps({
```

File: app/routes/docs.faq.md (hunk 3 above)
Comment (jracusin) [reply app/routes/docs.faq.md:71]: This will be moved from FAQ to tutorial page.

**Comment by Vidushi-GitHub** [comment]
@dakota002 I accidentally closed this branch, I was working on rebase!

**Comment by lpsinger** [comment]
That's alright, @Vidushi-GitHub. You closed it, but you didn't merge it. You could re-open it if you want to.
